### DISCUSSION TITLE
Is it possible to build a dynamic library with its dependencies statically linked for Windows?

### DISCUSSION THREAD

**Discussion by BewareMyPower** [microsoft/vcpkg#35733] [Q&A] [answered]
It's commonly seen on Linux that a dynamic library is built that all its dependencies are statically linked. For example, `A` depends on `B` and `C`, `b.a` and `c.a` will be built first, then `liba.so` will be built on them.

When I added a vcpkg port locally, it also worked well.

However, when I worked for Windows, when the triplet is `x64-windows-static`, if I don't remove the `bin/` directories, there will be a post-build error that indicates me to add the following lines:

```cmake
if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
endif()
```

Then the installed DLL file will be deleted while the corresponding imported LIB file is still retained.

I see when installing protobuf with the `x64-windows-static` triplet, protobuf with `x64-windows` triplet is also installed.

```
The following packages will be built and installed:
  * protobuf:x64-windows -> 3.21.12
    protobuf:x64-windows-static -> 3.21.12
```

It seems because it depends on itself as the host dependency: https://github.com/microsoft/vcpkg/blob/0e47c1985273129e4d0ee52ff73bed9125555de8/ports/protobuf/vcpkg.json#L8-L11

Could anyone explain how it works? Though I'm not sure if the protobuf.dll links statically on its dependencies.

**Comment by Osyotr** [comment, marked as answer]
> It's commonly seen on Linux that a dynamic library is built that all its dependencies are statically linked. 

You need to create a custom triplet that does something like this:
```cmake
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
if("${PORT}" STREQUAL "protobuf")
    set(VCPKG_LIBRARY_LINKAGE dynamic)
endif()
```
> I see when installing protobuf with the x64-windows-static triplet, protobuf with x64-windows triplet is also installed.

1) You need to build protobuf for host to get `protoc` that is used when cross-compiling protobuf
2) You asked for `x64-windows-static` but your host triplet is by default `x64-windows`. This is considered to be a cross-compilation.

If you want to save some time, you can make host triplet equal target triplet.

**Reply by BewareMyPower** [reply]
Thanks. I just thought of the solution for my issue before: https://github.com/microsoft/vcpkg/issues/27491#issuecomment-1299490075. It's acceptable to use a customized triplet.
